Branch remeshing

dngoldberg opened the pull request:
## What changes does this PR introduce?
This branch is a collection of updated and new files to implement the functionality described in Jordan et al (2017), "Ocean-forced ice-shelf thinning in a synchronously coupled ice–ocean model
". The main functionality is to allow "wetting" and "drying" of cells in the vertical, when using pkg/shelfice with SHELFICEMassStepping=true. The changes will only be seen with ALLOW_SHELFICE_REMESHING defined in SHELFICE_OPTIONS.h. The algorithm is described in detail in the paper above. There are also related functionalities, also described in the paper:

-- a change to allow realFWFlux when shelficeBoundaryLayer=.true., by carefully accounting for fluxes at the ice-ocean interface (requires shelficeUseFWFlux=.true.)
-- a change to the definition of uStar that partially addresses "stripes" in melt rate (requires #define SHI_USTAR_TOPDR)


## What is the current behaviour? 
None of the above functionalities implemented

## What is the new behaviour 
All of the above functionalities implement

## Does this PR introduce a breaking change? 
None. Have run ISOMIP verification to confirm.

## Other information:
Will update documentation for shelfice when it is introduced

jm-c commented:
@dngoldberg It would be good to have a set-up example (a "verification" experiment) to look at and check how this works. You might be able to take part of the current verification_other/shelfice_remeshing/ (in Contrib) if this is not too different.
Also, if there is any piece of documentation it would be great.

dngoldberg commented:
i can create a verification/shelfice_remeshing experiment, i will do this as part of the current pull request. Regarding documentation, we can produce this, but it will reference Shelfice extensively (for which there is currently no documentation).

christophernhill commented:
@jrscott should you create a new Section in manual (maybe somewhere between "Sea-ice" and "Biogeo")? Then @dngoldberg will have somewhere to put some doc pieces?  Maybe something like _Ice Shelf/Ice Sheet Interaction Packages_ ?

dngoldberg commented:
@christophernhill @jrscott that would be good as i will also need to write documentation for pkg/streamice. @jm-c i have committed a verification exercise shelfice_remeshing. I included it as a part of the same pull request, I hope this is what you intended.

edoddridge commented:
@dngoldberg it would be good to bring the documentation into the repo, even if it does reference the currently nonexistent shelfice docs. Otherwise your great package will never get documented and nobody will know how to use it.

I reckon you can create the file for this documentation yourself - it's nothing fancy just a text file with ".rst" as the extension. You'll need to modify `phys_pkgs.rst` to add your new section and file to the table of contents for the documentation.

Hopefully between the [contributing to the manual](http://mitgcm.readthedocs.io/en/latest/contributing/contributing.html#contributing-to-the-manual) section and the other files in the doc folder you can get a feel for how they work. If not, I'm happy help.

I agree with putting this new documentation between sea ice and biogeochemistry.

jrscott commented:
Indeed, I agree adding a new section between current 8.6 and 8.7 as @christophernhill suggests
"Shelf/Ice Sheet Interaction Packages" with shelfice to go as (the first, and only at this point) subsection. I can do a quick PR with this top-level change, thus @dngoldberg can write his own new .rst file for this package, and we can insert into phys_pkgs.rst.

In terms of what package doc should contain: this has not been finalized/written in stone, and I don't ever think one-shoe-fits-all will work. But I do think all namelist parms need to be documented.
See my current PR for chapter 3 parms (i.e. all the namelist parms in data, documented), even better steal the table format and adapt for your use!

Also suggest looking at some of the existing package doc to get a sense, but be reminded that this has not been checked/updated (for conformity, content etc.), on a long-term to-do list. In other words, significant bits of existing package docs need work. Below, see what I had written in MITgcm devel group meeting notes about "suggested package doc content". Don't hesitate to email me (and/or @edoddridge ) for help.

General Package documentation requirement:
-	What it does, how best used, appropriate for which setups etc. (also, e.g., how sensitive is a given setup to turning on this package? if applicable)
-	Equations/Description (scientific or otherwise), sample results (if appropriate)
-	Explanation of all namelist parms and diagnostics, guide to usage
-	Guide to implementation/subroutine calls (technical description)

edoddridge commented:
> I can do a quick PR with this top-level change, thus @dngoldberg can write his own new .rst file for this package, and we can insert into phys_pkgs.rst.

If @dngoldberg would prefer one of us to make the file, it would be better to push that change onto this branch rather than have it as a separate PR.

jm-c commented:
regarding pkg subsections, we could also extend current subsection 8.6 (sea ice packages)
to include all form of ice pkgs (seaice, shelfice, icefront, frazil & streamice); this way
we don't have to decide whether frazil go with seaice or with shelfice.

On Thu, Jul 19, 2018 at 04:44:12AM -0700, christophernhill wrote:
> @jrscott should you create a new Section in manual (maybe somewhere between "Sea-ice" and "Biogeo")? Then @dngoldberg will have somewhere to put some doc pieces?  Maybe something like _Ice Shelf/Ice Sheet Interaction Packages_ ?
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/124#issuecomment-406247913

dngoldberg commented:
Hello all. I have uploaded documentation for SHELFICE_REMESHING (not a separate package, but i guess deserving its own entry) to this PR. Unavoidably it does reference some aspects of SHELFICE. I have also modified phys_pkg.rst simply to have a way to display the page. there was a lot of discussion about how this should be structured -- i am pretty agnostic to the whole issue, and phys_pkg.rst can be changed as needed upon review. I think this completes the initial list of items asked for my J-M.

jm-c commented on model/src/forward_step.F:
```diff
@@ -388,6 +389,15 @@ SUBROUTINE FORWARD_STEP( iloop, myTime, myIter, myThid )
+        CALL TIMER_STOP('SHELFICE_REMESHING     [FORWARD_STEP]',myThid)
+#endif /* ALLOW_SHELFICE_REMESHING */
+#endif /* ALLOW_SHELFICE */
+
```
This call to SHELFICE_REMESHING is clearly at the wrong place, before DEBUG_ENTER !
should be pushed at least after AUTODIFF_INADMODE_UNSET call.

jm-c commented on model/src/upd_linear_phisurf.F:
```diff
@@ -0,0 +1,244 @@
+         ENDDO
+        ENDDO
+      ELSE
+        STOP 'INI_LINEAR_PHISURF: We should never reach this point!'
```
printed message after the STOP is not right.

jm-c commented on pkg/shelfice/SHELFICE.h:
```diff
@@ -33,6 +33,8 @@ C     SHELFICEadvDiffHeatFlux  :: use advective-diffusive heat flux into the
 C                                 flux, see Holland and Jenkins (1999),
 C                                 eq.21,22,26,31; def: F
+C     SHELFICErealFWflux       :: ensure vert advective flux at bdry uses top cell
+C                                 value rather than "boundary layer" value   F
```
what is "boundary layer value F" ?

jm-c commented on pkg/shelfice/SHELFICE.h:
```diff
@@ -52,6 +54,10 @@ C     shiPrandtl, shiSchmidt   :: constant Prandtl (13.8) and Schmidt (2432.0)
 C     shiKinVisc               :: constant kinetic viscosity used to compute
 C                                 gammaTurb (def: 1.95e-5)
+C     SHELFICERemeshFrequency  :: Frequency that size of etaN is checked to
+C                                 trigger remesh
```
the description of SHELFICERemeshFrequency is not very clear to me.

jm-c commented on pkg/shelfice/shelfice_forcing_surf.F:
```diff
@@ -1,3 +1,6 @@
+C $Header: /u/gcmpack/MITgcm_contrib/verification_other/shelfice_remeshing/code/shelfice_forcing_surf.F,v 1.5 2017/03/29 09:20:16 dgoldberg Exp $
+C $Name:  $
+
```
CVS Header and Name have been removed. So we are not going to add them back.

jm-c commented on pkg/shelfice/shelfice_remeshing.F:
```diff
@@ -0,0 +1,69 @@
+
+                
+
+
```
Only 6 consecutive blank lines ? why not 32 or 64 ?

jm-c commented on pkg/shelfice/shelfice_init_varia.F:
```diff
@@ -42,6 +45,8 @@ SUBROUTINE SHELFICE_INIT_VARIA( myThid )
           shelficeMass          (i,j,bi,bj) = 0. _d 0
+!          EFFMASS               (i,j,bi,bj) = 0. _d 0
+!          SeaLevelRestore       (i,j,bi,bj) = 0. _d 0 
          ENDDO
```
Why modifying this file for just adding 2 commented initialisation (of un-known variable) and CVS header and Name ?

dngoldberg commented:
@jm-c Thank you for looking over the proposed changes. I have verified that upd_linear_phisurf is extraneous. obcs_update_remesh does not need to be called, but I point out that for horizontal remeshing i.e. grounding line retreat, 2D masks are updated, and I think this might become necessary. I will respond to in-line comments when i push the updates you have asked for.

but first it seems some files need to be removed from the pull request. before i call any commands i cannot undo, can i please ask whether it is enough to call
git reset HEAD^ [/path/to/file]
before calling "commit" or whether additional/alternative steps are necessary?

dngoldberg commented:
@jm-c I have made the changes you requested. several files as you said should be part of separate P/R's as they are separate functionality so i have managed to remove that changes from this P/R, and have modified tag-index and the documentation (and the verification expt) accordingly, and removed files from P/R that should not be needed (e.g. obcs_update_remesh, upd_linear_phisurf, shelfice_init_varia).

jm-c commented:
I started to look at the remeshing code but few things are not very clear:

1) this src file: upd_masks_remesh.F 
 is sitting in model/src but it's empty when pkg/shelfice is not
 compiled ; it's only called from a pkg/shelfice S/R (shelfice_remeshing.F)
 and so a logical place would be (from my view) pkg/shelfice instead of 
 model/src ?

This makes it less easy to follow when new src code is not found at the
"logical" place (and in addition, comments are not very helpful, e.g.,
at the top of shelfice_update_ksurf_remesh.F :

      SUBROUTINE SHELFICE_UPDATE_KSURF_REMESH( myThid )
C     *============================================================*
C     | SUBROUTINE SHELFICE_INIT_FIXED
C     | o Routine to initialize SHELFICE parameters and variables.
C     *============================================================*
C     | Initialize SHELFICE parameters and variables.
C     *============================================================*

2) There are other coding problems (some related to missing some recent modification
of MITgcm src code) but I would like to start with the main part of the "remeshing",
code which seems to be in:
 pkg/shelfice/shelfice_update_masks_remesh.F
and specilly in the section just after (line 67) "C- Update etaN"
since the next sections:
  C- fill in the overlap (+ BARRIER):
and
  C--   Calculate lopping factor hFacC : Remove part outside of the domain
might not be needed if the remeshing code had been built on latest MITgcm
(post checkpoint66g, April 2017, from tag-index:
  - change S/R SHELFICE_UPDATE_MASKS to SHELFICE_INIT_DEPTHS (still called from
    ini_masks_etc.F for now) to update ocean upper boundary position according
    to ice-shelf topography (instead of directly updating hFacC).
)

a) criteria for remeshing:
 splitting of upper grid-cell (lines 75,76) or merging of 2 upper-most grid-cell
 (lines 99,100) only works if R_shelfice is exactly rF(kTopC) ; However, this
 is not generally the case, and I don't see any specific remeshing changes in 
 pkg/shelfice initialisation that would make it happens.
 I don't think it would be difficult to make it to work for the general case
 (i.e., any initial R_shelfice value), just need to check for hFac_surfC.

b) could you exlain what the second condition (lines 101,102) means ?
 it does not seem right to me.

c) In the splitting case, all the state variables in the new grid cell
 are taken fron the former -- too thick -- upper grid cell, and I think 
 it's OK (but one or 2 lines of comments in the src code would help),
 apart from missing gtNm1,gsNm1 and not compiling with ALLOW_ADAMSBASHFORTH_3

d) However the merging case has several issues:
  line 112: etaN(I,J,bi,bj) = etaN(I,J,bi,bj) +drF(K)
 and therefore when updating salt(:,:,k+1) to merge in salt from level k
 (lines 157-160), it should just be salt(:,:,k)*etaN instead of 
 salt(:,:,k)*(drF(k)+etaN) since etaN has been updated ?

e) And also in merging case, none of the update of uVel,vVel,guNm1,gvNm1
 is always right, because it all depends on neighbor column.

3) May be it would be faster to write down the remeshing algorithm,
 discuss the potential issues (e.g., with staggerTimestep and without)
 and agree on some solution before recoding this part ?

jm-c commented:
@dngoldberg I think for now it's better to focus on your new revision plan of
SHELFICE_UPDATE_MASKS_REMESH (your following email). But just to clarify:

> -- "change S/R SHELFICE_UPDATE_MASKS to SHELFICE_INIT_DEPTHS" -- i do not understand this comment
For some reason, indent in my email is modified when it appears on the GitHub PR page, 
so it's even more confusing.
I was quoting doc/tag-index, changed before checkpoint66g, April 2017.
Current version of ini_masks_etc.F no longer calls SHELFICE_UPDATE_MASKS in the middle of
this S/R but instead calls SHELFICE_INIT_DEPTHS at the top.
However, the remeshing S/R are more following the old ini_masks_etc.F/shelfice_update_masks.F
than the new way.

And regarding this:
> a) I hope this comment is limited to initialisation? It was decided, after each remeshing, that R_shelfice be set to the level of a cell interface to remove ambiguity. 
I have no problem with this.

> For instance: R_shelfice(i,j) might initially be (rF(kTopC(i,j))-8.71) simply because of the shape of the topography, which is fine. But when the ice shelf thins ~100m, it seems strange to always keep R_shelfice(i,j) always offset from rF(kTopC). 
I also agree with this.

> Moreover we found it easier to develop when this condition (that R_shelfice be on a vertical cell face) is ensured. Nor would it make sense to adjust R_shelfice by less than drF, because this could potentially lead to changing it each time step, which would be costly and unnecessary. 
This is not clear to me.

> As of now the user must ensure this property through initialisation files (of R_shelfice and etaInit) 
This is my problem, and as I wrote "it only works if R_shelfice is exactly rF(kTopC)" but there is
no piece of (remeshing) code that is doing it, no warning to user that they have to do it, so it's not great.

> -- but this can be changed. Can I suggest to modify eta and R_shelfice in the initialisation instead to ensure this property holds?
This would be one solution, but it can also be done by checking hFac_surfC, and might not be more
complicated that current remeshing code. Suggest to add this issue to "revision plan discussion".

jm-c commented:
I am adding new comments to this discussion:

1) To make the remeshing criteria work with any R_shelfice (and not only with
R_shelfice = rF(kTopC)), could try something like:

    k = kTopC
    IF ( hFac_surfC > SHELFICESplitThreshold ) THEN
C-  splitting surface grid cell (at kTopC) into level above (at k-1)
        adjust = rF(k-1) - R_shelfice
        R_shelfice = R_shelfice + adjust
        etaH = etaH - adjust
    ENDIF
    IF ( hFac_surfC < SHELFICEMergeThreshold ) THEN
C-  merging surface grid cell (at kTopC) into level below (at k+1)
        adjust = R_shelfice - rF(k+1)
        R_shelfice = R_shelfice - adjust
        etaH = etaH + adjust
    ENDIF

2) Regarding the Non-Linear FS variables at the start of forward_step.F
(i.e., where SHELFICE_REMESHING is called):
  hFacC,hFacW,hFacS                are consistent with etaHnm1
  hFac_surfC,hFac_surfW,hFac_surfS are consistent with etaH

Regarding the update of state variables (T,S,U,V):
 in order to conserve transport, uVel*hFacW and vVel*hFacS are the
 quantities that matter (we want to preserve the vertical integral);
 but to conserve Heat and Salt, it depends on staggerTimestep:
   if staggerTimestep=F, theta*hFacC and salt*hFacC matter.
 if staggerTimestep=T, need to consider
    theta*hFac_surfC and salt*hFac_surfC at the surface
    and theta*hFacC and salt*hFacC below

This is something to keep in mind for the merging case when adjusting
the state variables.

3) In current remeshing code (SHELFICE_REMESHING and all S/R called from
 there), all the geometric factor and mask are recomputed everywhere.
 An alternative would be to adjust some of them only where merging/splitting
 occurs and to recompute the other everywhere. This might be a little bit
 more efficient and may be simpler too. For instance, all the geometric
 factor (+mask,depth, ...) at grid-cell center could be adjusted where
 merge/split occurs while the one at Western and Southern edge would be 
 recomputed everywhere.
 Since the update of state variables (U,V,T,S) needs to be done
 where merge/split occurs, it would not necessary be a bad thing to have
 an update of geometric factors done at the same time.

4) I don't think we can call INTEGR_CONTINUITY from SHELFICE_REMESHING,
 because there is much more than computing wVel that is done there (e.g.,
 etaN is stepped forward since exactConserv=T to use nonlinFreeSurf > 0).
 Should instead just take the part that computed wVel.


On Tue, Sep 04, 2018 at 04:21:44PM +0000, dngoldberg wrote:
> @jm-c Here is an overview of what the SHELFICE_UPDATE_MASKS_REMESH is meant to do. And I have added changes not yet implemented in code (such as an extra step to update data at u- and v-points) following from the discussion above.
> 
> REMESHING ALGORITHM
> 
> UPD_MASKS_REMESH
>     SHELFICE_UPDATE_MASKS_REMESH
>       1. Check all columns to determine if top cells are too "thick" -- hFacC larger than imposed value
>          a) Further ensure that splitting cell will not result in merging in the next remeshing process
>       2. In these columns:
>          a) Raise R_shelfice by drF (kTopC-1) (and update Ro_surf accordingly)
>          b) Lower Eta by drF (kTopC-1)
>          c) set all *cell-centered* data in new cell to values just below
>       3. Check all columns to determine if top cells are too "thin" -- hFacC smaller than imposed value
>          a) Further check that merging cell will  not result in splitting in the next remeshing process
>       4. In these colums: 
>          a) Lower R_shelfice by drF (kTopC) (and update Ro_surf accordingly)
>          b) Raise Eta by drF (kTopC)
>          c) set all *cell-centered* data in new top-level cell via a weighted average of previous values
>       5. Save temporary copy of maskS/W (maskS/W_nm1), hFacS/W (hFacS/W_nm1)
>     	  
> 	.... logic to update hFacC/S/W, maskC/S/W and hFacC/S/W
> 
>     SHELFICE_UPDATE_UV_REMESH (new S/R)
> 
>       1. Find faces where maskS/W=1 and maskS/W_nm1=0. Set U/V and gU/Vnm1 to values in cell just below.
>       2. Find faces where maskS/W=0 and maskS/W_nm1=1. 
> 	     a) if maskS/W_nm1 = maskS/W = 1 in face immediately below, then update face-centered data in face immediately below in a manner that conserves volume flux.
> 		 b) otherwise, do not update.
> 
> ... REST OF SHELFICE_REMESHING
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/124#issuecomment-418430912

dngoldberg commented:
> I am adding new comments to this discussion: 1) To make the remeshing criteria work with any R_shelfice (and not only with R_shelfice = rF(kTopC)), could try something like: k = kTopC IF ( hFac_surfC > SHELFICESplitThreshold ) THEN C- splitting surface grid cell (at kTopC) into level above (at k-1) adjust = rF(k-1) - R_shelfice R_shelfice = R_shelfice + adjust etaH = etaH - adjust ENDIF IF ( hFac_surfC < SHELFICEMergeThreshold ) THEN C- merging surface grid cell (at kTopC) into level below (at k+1) adjust = R_shelfice - rF(k+1) R_shelfice = R_shelfice - adjust etaH = etaH + adjust ENDIF 2) Regarding the Non-Linear FS variables at the start of forward_step.F (i.e., where SHELFICE_REMESHING is called): hFacC,hFacW,hFacS are consistent with etaHnm1 hFac_surfC,hFac_surfW,hFac_surfS are consistent with etaH Regarding the update of state variables (T,S,U,V): in order to conserve transport, uVel*hFacW and vVel*hFacS are the quantities that matter (we want to preserve the vertical integral); but to conserve Heat and Salt, it depends on staggerTimestep: if staggerTimestep=F, theta*hFacC and salt*hFacC matter. if staggerTimestep=T, need to consider theta*hFac_surfC and salt*hFac_surfC at the surface and theta*hFacC and salt*hFacC below This is something to keep in mind for the merging case when adjusting the state variables. 3) In current remeshing code (SHELFICE_REMESHING and all S/R called from there), all the geometric factor and mask are recomputed everywhere. An alternative would be to adjust some of them only where merging/splitting occurs and to recompute the other everywhere. This might be a little bit more efficient and may be simpler too. For instance, all the geometric factor (+mask,depth, ...) at grid-cell center could be adjusted where merge/split occurs while the one at Western and Southern edge would be recomputed everywhere. Since the update of state variables (U,V,T,S) needs to be done where merge/split occurs, it would not necessary be a bad thing to have an update of geometric factors done at the same time. 4) I don't think we can call INTEGR_CONTINUITY from SHELFICE_REMESHING, because there is much more than computing wVel that is done there (e.g., etaN is stepped forward since exactConserv=T to use nonlinFreeSurf > 0). Should instead just take the part that computed wVel.
> […](#)
> On Tue, Sep 04, 2018 at 04:21:44PM +0000, dngoldberg wrote: @jm-c Here is an overview of what the SHELFICE_UPDATE_MASKS_REMESH is meant to do. And I have added changes not yet implemented in code (such as an extra step to update data at u- and v-points) following from the discussion above. REMESHING ALGORITHM UPD_MASKS_REMESH SHELFICE_UPDATE_MASKS_REMESH 1. Check all columns to determine if top cells are too "thick" -- hFacC larger than imposed value a) Further ensure that splitting cell will not result in merging in the next remeshing process 2. In these columns: a) Raise R_shelfice by drF (kTopC-1) (and update Ro_surf accordingly) b) Lower Eta by drF (kTopC-1) c) set all *cell-centered* data in new cell to values just below 3. Check all columns to determine if top cells are too "thin" -- hFacC smaller than imposed value a) Further check that merging cell will not result in splitting in the next remeshing process 4. In these colums: a) Lower R_shelfice by drF (kTopC) (and update Ro_surf accordingly) b) Raise Eta by drF (kTopC) c) set all *cell-centered* data in new top-level cell via a weighted average of previous values 5. Save temporary copy of maskS/W (maskS/W_nm1), hFacS/W (hFacS/W_nm1) .... logic to update hFacC/S/W, maskC/S/W and hFacC/S/W SHELFICE_UPDATE_UV_REMESH (new S/R) 1. Find faces where maskS/W=1 and maskS/W_nm1=0. Set U/V and gU/Vnm1 to values in cell just below. 2. Find faces where maskS/W=0 and maskS/W_nm1=1. a) if maskS/W_nm1 = maskS/W = 1 in face immediately below, then update face-centered data in face immediately below in a manner that conserves volume flux. b) otherwise, do not update. ... REST OF SHELFICE_REMESHING -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#124 (comment)](https://github.com/MITgcm/MITgcm/pull/124#issuecomment-418430912)

dngoldberg commented:
@jm-c there is something i forgot about when i wrote the pseudo-algorithm before, and need to add to the discussion. Rmin_surf is generally set in ini_nlfs_vars.F, but will need to change as Ro_surf etc changes. This was in the contrib code written by @JimRowanJordan but when i updated this i omitted this functionality by mistake. Therefore i update the pseudo algorithm to reflect this change, and the discussion above. Please let me know if you disagree with changes Rmin_surf.

REMESHING ALGORITHM


	UPD_MASKS_REMESH
		SHELFICE_UPDATE_MASKS_REMESH
			1. Check all columns to determine if top cells are too "thick" -- hFacC larger than imposed value
				a) Further ensure that splitting cell will not result in merging in the next remeshing process
			2. In these columns:
				a) Raise R_shelfice by drF (kTopC-1) (and update Ro_surf accordingly)
				b) Lower Eta by drF (kTopC-1)
				c) set all *cell-centered* data in new cell to values just below
                                d) update hFacC in cells (kTopC, kTopC-1) accordingly
			3. Check all columns to determine if top cells are too "thin" -- hFacC smaller than imposed value
				a) Further check that merging cell will  not result in splitting in the next remeshing process
			4. In these colums: 
				a) Lower R_shelfice by drF (kTopC) (and update Ro_surf accordingly?)
				b) Raise Eta by drF (kTopC)
				c) set all *cell-centered* data in new top-level cell via a weighted average of 
previous values
                                d) update hFacC in cells (kTopC, kTopC-1) accordingly
			5. Save temporary copy of maskS/W (maskS/W_nm1), hFacS/W (hFacS/W_nm1)
    	  
		.... logic to update hFacS/W, maskS/W and hFacS/W
                .... logic to update Rmin_surf

		SHELFICE_UPDATE_UV_REMESH (new S/R)

			1. Find faces where maskS/W=1 and maskS/W_nm1=0. Set U/V and gU/Vnm1 to values in cell just below.
			2. Find faces where maskS/W=0 and maskS/W_nm1=1. 
				a) if maskS/W_nm1 = maskS/W = 1 in face immediately below, then update face-centered data in face immediately below in a manner that conserves volume flux.
				b) otherwise, do not update.

	... REST OF SHELFICE_REMESHING

dngoldberg commented:
@jm-c  I have now made and tested changes to the shelfice_remeshing branch. These changes address most (if not all?) of the issues you bring up. 

-- I removed model/src/upd_masks_remesh. There are no new source files outside of pkg/shelfice.
-- The remeshing process now allows for R_shelfice to not initially be positioned at a vertical cell face.
-- I have addressed the bug which merged salt and temperature incorrectly when merging cells.
-- I have modified the code which updated dynamic variables at U- and V-points. This is done by first allowing new geometric factors to be calculated (using the same algorithm as ini_masks_etc) and then updating only where masks have changed, as described in the algorithm in my previous post.
-- I have attempted to minimize the reinitialisation "from scratch" of cell-centered masks and geom factors. the algorithm used to find u- and v- geom factors requires hfacc=h0facc, so only h0facc and maskC are updated incrementally; hfacC and hfac_surfC are sorted by calls to calc_surf_dr and upd_surf_dr. 2D index arrays (ksurfC etc) are also updated incrementally
-- i have stripped out from integr_continuity only the commands needed (see SHELFICE_INTEGR_CONT_REMESH)
-- i include a flag to merge theta/salt differently based on staggered timestepping
-- i update Rminsurf -- otherwise there are extensive SURF_ADJ errors.
-- the only other recent updates are to the verif experiment to better facilitate 2+2 testing (remeshing occurs mid-simulation rather than the 1st timestep, so this can test pickups more extensively)
-- these changes were checked against the previous commits on this branch and were identical where they should be (i.e. where there is no merging, only splitting)

@edoddridge @jrscott I have not yet updated the documentation, as this should be done only after the code is finalised, and possibly s/b a separate PR.

jm-c commented:
@dngoldberg  I looked at the recent version, and it looks better.

1) Regarding the main remeshing part:
 We still have a lot of geometric factor which are recomputed from scratch 
 (at least for Western and Southern location), partly in shelfice_remesh_uvmasks.F ;
 I think it should be possible to just adjust them where they need to be changed;
 But we could agree that this could be part of a future "improvement" (specially
 if the "recomputation from scratch" helps for the horizontal remeshing).
 What is your opinion on this ?

2) even if recompute all the W & S geometric factor, I would suggest to clean-up
 a bit the current code and NOT recompute every-where any of the grid-cell centered
 geometric factor:  recip_Rcol, hFacC, maskC, kSurfC, kTopC, .... 
 Also, we probably don't need to save maskW,maskS in maskW_temp, maskS_temp,
 as long as the save hFacW,hFacS into hFacW_temp,hFacS_temp (but names here
 are not so great, may be "_old" or "_sav" would be better).

3) As I wrote earlier (Sep 04):
  > hFacC,hFacW,hFacS                are consistent with etaHnm1
  > hFac_surfC,hFac_surfW,hFac_surfS are consistent with etaH
  So the current sequence of call:
    CALL CALC_SURF_DR( etaH ,
    CALL CALC_SURF_DR( etaH ,
    CALL UPDATE_SURF_DR( .FALSE., 
 does not look exactly right: I think etaHnm1 should be adjusted also,
 in shelfice_update_masks_remesh.F, the same way etaH is: 
    etaH(I,J,bi,bj) = etaH(I,J,bi,bj)- adjust
      ...
    etaH(I,J,bi,bj) = etaH(I,J,bi,bj)+ adjust
 and then we should have something like:
    CALL CALC_SURF_DR( etaHnm1 ,
    CALL UPDATE_SURF_DR( .TRUE.,
    CALL -- recompute vertical velocity -- 
    CALL CALC_SURF_DR( etaH ,
 This is closer to what is done in initialise_varia.F
 Note that I don't think it matter when the vertical velocity is recomputed,
 before 2nd CALC_SURF_DR call or after, but would be nice to keep same sequence
 of calls than in initialise_varia.F.

4) regarding new S/R names, I think it would help to have names starting with 
  "SHELFICE_REMESH_", except for the main one that can be kept as "SHELFICE_REMESHING"
   I could suggest some shorter name like these:
   SHELFICE_UPDATE_MASKS_REMESH   --> SHELFICE_REMESH_C_MASK
   SHELFICE_REMESH_UVMASKS        --> SHELFICE_REMESH_UVMASKS
   SHELFICE_REMESH_UPDATE_RMINSURF -> SHELFICE_REMESH_RMINSURF
   SHELFICE_INTEGR_CONT_REMESH     -> SHELFICE_REMESH_CALC_W
 but you should pick names that you like.
 Also it would be simpler to call all of them from SHELFICE_REMESHING,
 (instead of have some of them called from  SHELFICE_UPDATE_MASKS_REMESH);
 but this could be left for later (it it's not obvious).

dngoldberg commented:
hi @jm-c thank you for looking at it. to answer your questions:

1. i think here you are talking mainly about western and southern factors. my opinion is that in the interest of moving this forward i would prefer to calculate western/southern from scratch for the moment. I made quite a few attempts at making this computation "incremental" and every attempt met with failure -- it is very difficult to do cleanly (i.e. in a single pass) due to their dependency on 2 adjacent values. I think that, at least, all west/south factors are contained to the shelfice_remesh_uvmask routine -- making a clear target for future attempts. I am happy (as you suggest) to reduce recomputation in shelfice_update_masks_remesh.F.

Can i ask what is the object here in reducing recomputation? Is it to reduce processing time (which is small) or to minimise the chance of changes of initialisation S/R's leading to incompatibilities with the remeshing code?

2. I think some of these can be easily changed, e.g. recip_rcol, and i thikn ksurfc and ktopc. Some not. An exception is that I could see no other way than updating hfacc (everywhere) to h0facc and then allowing update_surf_dr to set the correct hfacc. As I say above, i am not opposed to revisiting this, but the complexity is quite high and it might be good to leave it for the future.

I like the idea of using maskW/S rather than hFacW/S because it is binary -- is it an issue of memory? Or is there an issue i am not seeing?

currently i am running the MISOMIP experiment and it is exposing some bugs with what I have written, not picked up on previously.  if i do not recompute rSurfW/S then the hfacW/S is not recomputed correctly when  useMin4hFacEdges=.false. This represents yet another field that, if done incrementally, would be difficult due to the dependency on adjacent values. do you have any suggestion here?

3. The complexity here was too great for me -- it is not that i question you but i do struggle to see what is wrong with calling calc_surf_dr twice. I will try as you suggest by looking at initialise_varia.F, but i may struggle.

4. no strong opinion on S/R names.

jm-c commented:
Hi @dngoldberg,

Regarding point (1), I was a bit concerned about efficiency (for instance, thin-wall
will be read-in from file every remeshing time-step) since initialisation code
is not intended to be efficient (only done once). Also there could be some limitations
if other "remeshing" (wetting and drying in coastal shallow area) were to be
implemented to work also with remeshing under ice-shelf.
But we can leave this to a future improvement (beyond this PR), I agree.

So, In summary: 
 - all cell and column centered quantities: adjusted/changed locally.
 - all western & southern edges arrays: recomputed form scratch (as you did).

Regarding the storage of maskW,maskS, yes, memory footprint (and efficiency) 
would be the main reason. 

Regarding point (3), It would be good if you could try my suggestion, and see
how it works (should work better ?).

On Thu, Nov 15, 2018 at 03:10:00PM -0800, dngoldberg wrote:
> hi @jm-c thank you for looking at it. to answer your questions:
> 
> 1. i think here you are talking mainly about western and southern factors. my opinion is that in the interest of moving this forward i would prefer to calculate western/southern from scratch for the moment. I made quite a few attempts at making this computation "incremental" and every attempt met with failure -- it is very difficult to do cleanly (i.e. in a single pass) due to their dependency on 2 adjacent values. I think that, at least, all west/south factors are contained to the shelfice_remesh_uvmask routine -- making a clear target for future attempts. I am happy (as you suggest) to reduce recomputation in shelfice_update_masks_remesh.F.
> 
> Can i ask what is the object here in reducing recomputation? Is it to reduce processing time (which is small) or to minimise the chance of changes of initialisation S/R's leading to incompatibilities with the remeshing code?
> 
> 2. I think some of these can be easily changed, e.g. recip_rcol, and i thikn ksurfc and ktopc. Some not. An exception is that I could see no other way than updating hfacc (everywhere) to h0facc and then allowing update_surf_dr to set the correct hfacc. As I say above, i am not opposed to revisiting this, but the complexity is quite high and it might be good to leave it for the future.
> 
> I like the idea of using maskW/S rather than hFacW/S because it is binary -- is it an issue of memory? Or is there an issue i am not seeing?
> 
> currently i am running the MISOMIP experiment and it is exposing some bugs with what I have written, not picked up on previously.  if i do not recompute rSurfW/S then the hfacW/S is not recomputed correctly when  useMin4hFacEdges=.false. This represents yet another field that, if done incrementally, would be difficult due to the dependency on adjacent values. do you have any suggestion here?
> 
> 3. The complexity here was too great for me -- it is not that i question you but i do struggle to see what is wrong with calling calc_surf_dr twice. I will try as you suggest by looking at initialise_varia.F, but i may struggle.
> 
> 4. no strong opinion on S/R names.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/124#issuecomment-439223711

christophernhill commented:
@dngoldberg did you get a chance to look at @jm-c latest comments?

dngoldberg commented:
hi @christophernhill @jm-c yes i did, while i am a bit skeptical i can remove even more initialisation code than i already have, i have been attempting to run the MISOMIP experiment (https://www.geosci-model-dev.net/9/2471/2016/gmd-9-2471-2016.pdf) using the updated code as the deadline is approaching and the model has been blowing up and i cannot tell if this is due to a bug -- so between this and everything else i have to do i have not been able to focus on this P/R.

jm-c commented:
@dngoldberg Did you had chance to try my point-3 suggestion from Nov 15 ? I expect this modification to
change the results (for a good reason) so it would be good if you can try it. The other points are not 
so important (coding preference) since they will not affect results.

dngoldberg commented:
apologies @jm-c  i have been focusing on the MISOMIP intercomparison (for which this code is being used), so i have not yet looked into this. I think i understand what you are suggesting re point 3 and I am going to try it -- and i will see what happens..

dngoldberg commented:
Hi @jm-c 

1. I have tried your suggestion re: point 3 abive. It changes the code quantitatively but does not fail. The approach I have taken is to call CALC_SURF_DR the second time just after UPDATE_SURF_DR; doing it after the vertical velocity recalc would mean moving it to a different S/R and as you say this is not important.

2. I have changed the S/R names as you suggested in your last review.

3. I have not removed any of the recomputation of cell centered factors i already have. As I said before I believe they are necessary. E.g. recip_RCol, which you said could be left alone without recomputing, changes (in an nonlinear way) when Rosurf changes. These and the other cell-centered arrays which i recomputed, I could not see any other way. I prefer to use maskS and maskW to check whether a cell has been "dried", as this seems more robust, but I am willing to debate this. I have stopped using the "_temp" naming and use "_old" instead.

3. In extensive testing I discovered that rsurfS/W (which is based on Ro_surf) was not previously being updated. I have added code to do this as well.

dngoldberg commented:
Hi @jm-c (and @christophernhill) I have thought more about the issue of the modified calls to CALC/UPDATE_SURF_DR.

Apologies if below only refers to one specific mode of running (e.g. staggered timestep, nonlin free surface, no rstar), it's just that it is the only mode i am familiar with.

Without remeshing, at the start of each time step, as I understand it:
1) hFac_surfC/S/W corresponds to etaH
2) hFac_surfNm1C/S/W corresponds to etaHnm1 (i think)
3) hfacC/S/W has been updated with hFac_surfC/S/W i.e. their "top" values reflect etaH

Remeshing occurs "outside time" therefore the above should be true after remeshing occurs, but with the modified definitions of etaH (and etaHnm1).

You have suggested the ordering:

CALL CALC_SURF_DR( etaHnm1, myTime, myIter, myThid )
CALL UPDATE_SURF_DR(.TRUE., myTime, myIter, myThid)
CALL CALC_SURF_DR( etaH, myTime, myIter, myThid )

but i do not think they will achieve this. (1) and (2) above will be achieved, but (3) will not. i suggest a reversing of the last two calls:

CALL CALC_SURF_DR( etaHnm1, myTime, myIter, myThid )
CALL CALC_SURF_DR( etaH, myTime, myIter, myThid )
CALL UPDATE_SURF_DR(.TRUE., myTime, myIter, myThid)

what do you think? 

after this issue there is only the issue of which fields should be recomputed, and where.

thanks
dan

christophernhill commented:
@jm-c to get back to Dan

jm-c commented:
@dngoldberg looking at your latest changes from June 06 regarding reading pickup:
https://github.com/MITgcm/MITgcm/pull/124/commits/8deb5a7b13bfd1e16da7f633987cb0e2a905a79b
I think I understand the point but the coding is not exactly right (we don't need to read a pickup
form S/R SHELFICE_INI_DEPTH when ALLOW_SHELFICE_REMESHING is undef). Will clean up this.

dngoldberg commented:
Hi @jm-c Sorry have been slow to respond to this. No issues with the rearrangement of the verification_other experiment, i am agnostic on this point (assuming i understand it). 

Your point about pickup -- even if mass_stepping is set to false, initialisation that is not very, very carefully done can result in changes in eta, and potentially trigger remeshing. Sea level change (e.g. due to obcs velocities that are not corrected) can lead to this as well... So in the spirit of robustness, this is why i left this pickup in, as R_shelfice should be a pickup field. (if shelfice_mass_stepping is false, then mass should of course not be a pickup field -- and i can't recall whether i ensured this.)

jm-c commented:
@dngoldberg
I would like to make (3) significant changes to this PR:

1) replace the verification experiment "shelfice_remeshing" with a simpler one
  (suggest: shelfice_2d_remesh), with similar resolution but:
 a) no pkg/streamice but instead prescribed ice-mass tendency (SHELFICEMassDynTendFile),
    really 2-D (y-z) set-up (currently, because of pkg/streamice, Nx=3), with Coriolis on
    (so that uVel <>0). 
 b) no SHELFICEboundaryLayer so that it's just more straightforward to get real fresh-water in.
 c) un-realistically large ice-mass tendency so that several remeshing happen during
   the short testreport run (20 time-step), and specially remeshing of both types (merging 
   and splitting). The problem this current shelfice_remeshing experiment it that remeshing 
   occurs only once during the short 20.iter run, a splitting one, so that the merging case 
   was never tested.
 d) current "shelfice_remeshing" experiment in this PR does not have to be lost but
   could be used to update the same name set-up in verification_other or could move there 
   with a different name.
 
2) move away from this PR the changes in shelfice_thermodynamics.F ( SHI_USTAR_TOPDR
   and SHELFICErealFWflux features) so that this PR will be only dealing with remeshing.
   Can discuss (on the side) what to do with each of them but since they both relate
   to SHELFICEboundaryLayer and the simplified test experiment "shelfice_2d_remesh" I have
   been playing with does not use SHELFICEboundaryLayer, it should not interfere with this PR.

3) more cleaning of the remeshing code to come. But should be easier to check that
  things are right with a short & simple test experiment.

Let me know what you think.

jm-c commented:
After consulting @dngoldberg, I added my simplified remeshing test experiment. I also re-worked 
the 2 new features (formerly: SHI_USTAR_TOPDR and SHELFICErealFWflux, now both run-time
switches: SHI_withBL_realFWflux and SHI_withBL_uStarTopDz) and updated shelfice_remeshing exp.

I think it's better to leave further changes/simplification of the remeshing code (my 3rd point in previous 
comment) to an other PR, so that, if the latest changes are OK with you, will be able to merge this PR
after I move away the shelfice_remeshing experiment.

jm-c commented:
@dngoldberg So, are you OK with me moving away (to verification_other) the current
"shelfice_remeshing" experiment (and just keeping the simpler "shelfice_2d_remesh") ?

jm-c commented:
@dngoldberg I made a new PR for verification_other repos with a copy of your original
"shelfice_remeshing" experiment set-up. So I am ready to remove it from this PR and after that
to merge it.
